src/main/kotlin/com/openlattice/rhizome/core/service/ContinuousRepeatingTaskService.kt

```diff
@@ -0,0 +1,146 @@
+package com.openlattice.rhizome.core.service
+
+/*
+ * Copyright (C) 2020. OpenLattice, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * You can contact the owner of the copyright at support@openlattice.com
+ *
+ *
+ */
+
+import com.google.common.util.concurrent.ListeningExecutorService
+import com.hazelcast.core.HazelcastInstance
+import org.slf4j.Logger
+import java.time.Instant
+import java.util.concurrent.Semaphore
+import java.util.concurrent.TimeUnit
+
+internal const val DEFAULT_BATCH_TIMEOUT_MILLIS = 120_000L
+
+abstract class ContinuousRepeatingTaskService<T: Any>(
+        private val executor: ListeningExecutorService,
+        hazelcastInstance: HazelcastInstance,
+        private val logger: Logger,
+        queueName: String,
+        lockingMapName: String,
+        parallelism: Int,
+        private val fillChunkSize: Int = 0,
+        private val batchTimeout: Long = DEFAULT_BATCH_TIMEOUT_MILLIS
+) {
+    private val locks = hazelcastInstance.getMap<T, Long>(lockingMapName)
+    private val candidates = hazelcastInstance.getQueue<T>(queueName)
+    private val limiter = Semaphore(parallelism)
+
+    private val enqueuer = if ( enqueuerEnabledCheck() ) {
+        executor.submit {
+            while (true) {
+                try {
+                    sourceSequence()
+                            .filter {
+                                val expiration = lockOrGetExpiration(it)
+                                logger.debug(
+                                        "Considering candidate {} with expiration {} at {}",
+                                        it,
+                                        expiration,
+                                        Instant.now().toEpochMilli()
+                                )
+                                if (expiration != null && Instant.now().toEpochMilli() >= expiration) {
+                                    logger.info("Refreshing expiration for {}", it)
+                                    //Assume original lock holder died, probably somewhat unsafe
```

Comment:
So, the more I think about this logic the less comfortable I am with it. The logic I think we want is:

- Don't put things in the queue that are already in the queue
- If something is taken out of the queue and processed successfully, allow it to be put in the queue again.
- If something is taken out of the queue and not processed within a timeout, allow it to be put in the queue again.

This code (and I know it's not yours, you're adapting the existing pattern) does something different:
- If something is put in the queue for the first time, allow it and mark it.
- If the same entry is submitted within a timeout value, extend the timeout and do not put it in the queue.
- If the timeout expires, allow the object to be put in the queue even if it's still there from before.
This has the pathological case of a consumer crashes/eats an entry from the queue that's continuously being fed from the producer is never actually processed.

I think the impl we want is something like this:
- create an enum of queue state (NOT_QUEUED, QUEUED, PROCESSING)
- When putting something in the queue, put the key in the map with value QUEUED
- When taking something out of the queue, mark it with a timeout (value PROCESSING with timeout now + TTL)
- When finished processing an item from the queue, remove it from the map completely (allows resubmission).

Reply:
Yep, that makes sense. I'll refactor